src/runtime_src/core/pcie/emulation/hw_em/alveo_shim/shim.cxx

```diff
@@ -211,18 +212,19 @@ namespace xclhwemhal2 {
             if(std::find(parsedMsgs.begin(), parsedMsgs.end(), line) == parsedMsgs.end()) {
               logMessage(line);
               parsedMsgs.push_back(line);
-              if (!matchString.compare("Exiting xsim") || !matchString.compare("FATAL_ERROR")) {
+              if (!matchString.compare("Exiting xsim") || !matchString.compare("ERROR")) {
                   std::cout << "SIMULATION EXITED" << std::endl;
-                  this->xclClose();                                               // Let's have a proper clean if xsim is NOT running
-                  exit(0);                                                        // It's a clean exit only.
+                  lstatus = -1;
+                  this->xclClose(true);                                               // Let's have a proper clean if xsim is NOT running
+                 // exit(0);                                                        // It's a clean exit only.
               }
```

Comment:
This code is too deeply nested.  Keep the interesting code to the left, in other words, use early-exit.

Reply:
Agreed. The exit code check happens at first of the loop.